ARCs/arc-0047.md

```diff
@@ -0,0 +1,169 @@
+
+* An `address` value must be a valid Algorand address
+* A `uint64`, `application`, or `asset` value must be a valid unsigned 64-bit integer
+
+### Unique Identification
+
+To enable unique identification of a description, clients **MUST** calculate the SHA256 hash of the JSON description canonicalized in accordance with RFC 8785.
+
+## Rationale
+
+This provides a way for frontends to clearly display to the user what is being signed when signing a logic signature. 
+
+Template variables must be  an immediate argument otherwise a string variable could specify the opcode in the program, which could have unintended and unclear consequences. 
+
+`TMPL_` prefix is used to align with existing template variable tooling.
+
+Hashing canonicalized JSON is useful for ensuring clients, such as wallets, can create a allowlist of templated logic signatures. 
+
+## Backwards Compatibility
+N/A
+
+## Test Cases
+N/A
+
+## Reference Implementation
+
+### Description
+```json
+{
+    "name": "Address Opt-In",
+    "description": "This program allows a given address to opt in the signer to any asset provided it's approved by the associated application",
+    "program": "I3ByYWdtYSB2ZXJzaW9uIDgKI2RlZmluZSBNYXN0ZXJBcHBDYWxsIGxvYWQgMAoKLy8gU2F2ZSBNYXN0ZXJBcHBDYWxsCnR4biBHcm91cEluZGV4CmludCAxCisKc3RvcmUgMAoKLy8gVmVyaWZ5IGFtb3VudCBpcyAwCnR4biBBc3NldEFtb3VudAppbnQgMAo9PQphc3NlcnQKCi8vIFZlcmlmeSBzZW5kZXIgPT0gcmVjZWl2ZXIKdHhuIEFzc2V0UmVjZWl2ZXIKdHhuIFNlbmRlcgo9PQphc3NlcnQKCi8vIFZlcmlmeSBmZWUgaXMgMCAoY292ZXJlZCBieSBzZW5kZXIpCnR4biBGZWUKaW50IDAKPT0KYXNzZXJ0CgovLyBWZXJpZnkgYXNzZXRDbG9zZVRvIGlzIG5vdCBzZXQKdHhuIEFzc2V0Q2xvc2VUbwpnbG9iYWwgWmVyb0FkZHJlc3MKPT0KYXNzZXJ0CgovLyBWZXJpZnkgY2FsbGVkIGF0b21pY2FsbHkgd2l0aCBtYXN0ZXIgYXBwCk1hc3RlckFwcENhbGwKZ3R4bnMgQXBwbGljYXRpb25JRAppbnQgVE1QTF9ERUxFR0FURURfT1BUSU5fQVBQX0lECj09CmFzc2VydAoKLy8gVmVyaWZ5IHRoZSBjb3JyZWN0IG1ldGhvZCBpcyBiZWluZyBjYWxsZWQKTWFzdGVyQXBwQ2FsbApndHhuc2EgQXBwbGljYXRpb25BcmdzIDAKbWV0aG9kICJhZGRyZXNzT3B0SW4ocGF5LGF4ZmVyKXZvaWQiCj09CmFzc2VydAoKLy8gVmVyaWZ5IHRoZSBzZW5kZXIgaXMgdGhlIGNvcnJlY3QgYWRkcmVzcwpNYXN0ZXJBcHBDYWxsCmd0eG5zIFNlbmRlcgphZGRyIFRNUExfQVVUSE9SSVpFRF9BRERSRVNTCj09",
+    "variables": {
+        [
+            {
+                "variable": "TMPL_DELEGATED_OPTIN_APP_ID",
+                "name": "Delegated opt-in app ID",
+                "type": "application",
+                "description": "The ID of the application that will be used for verifying opt ins"
+            },
+            {
+                "variable": "TMPL_AUTHORIZED_ADDRESS",
+                "name": "Aurhotized address",
+                "type": "address",
+                "description": "The address that will be allowed to opt the signer into assets provided it's approved by the associated application"
+            }
+        ]
+    }
+}
+```
+
+### String Variables
+
+
+#### Invalid: Partial Argument
+```
+#pragma version 9
+byte "Hello, TMPL_NAME"
+```
+
+This is not valid because `TMPL_NAME` is not the full immediate argument. 
```

Comment:
This example drives home that it's a big ask for the wallet to parse TEAL.  According to these rules, it needs to accept:
`byte TMPL_NAME; int 7`, `byteblock "hello\" " TEMPL_NAME` for example.

I think we need to spec out the exact local rules that would allow a program to accept. I admit that's not easy.

Perhaps we should restrict things much more strongly. Only things that can be be properly escaped by the wallet, and result in a single `byte X` or `int X` instruction, for example

Reply:
Or probably `/^\s*(byte|int) TMPL_` to account for indented lines.

Reply:
I'd allow `pushint` and `pushbyte` as well.  I think there are good reasons to use them in templates, so that the inserted bytes end up exactly where you expect them.

(`int` and `byte` do some funny business around optimization, and moving the constants into constant blocks.)